sw/host/opentitanlib/src/transport/hyperdebug/i2c.rs

```diff
@@ -0,0 +1,234 @@
+            let mut req = CmdTransferShort::new();
+            req.port = self.bus_idx | (((wbuf.len() & 0x0F00) >> 4) as u8);
+            req.addr = addr;
+            req.write_count = (wbuf.len() & 0x00FF) as u8;
+            req.read_count = rbuf.len() as u8;
+            let databytes = std::cmp::min(USB_MAX_SIZE - 4, wbuf.len());
+            req.data[0..databytes].clone_from_slice(&wbuf[0..databytes]);
+            self.usb_write_bulk(&req.as_bytes()[0..4 + databytes])?;
+            databytes
+        } else {
+            // Long format header
+            let mut req = CmdTransferLong::new();
+            req.port = self.bus_idx | (((wbuf.len() & 0x0F00) >> 4) as u8);
+            req.addr = addr;
+            req.write_count = (wbuf.len() & 0x00FF) as u8;
+            req.read_count = (rbuf.len() & 0x007F) as u8;
+            req.read_count1 = (rbuf.len() >> 7) as u8;
+            let databytes = std::cmp::min(USB_MAX_SIZE - 6, wbuf.len());
+            req.data[0..databytes].clone_from_slice(&wbuf[0..databytes]);
+            self.usb_write_bulk(&req.as_bytes()[0..6 + databytes])?;
+            databytes
+        };
+
+        // Transmit any more data without further header.
+        while index < wbuf.len() {
+            let databytes = std::cmp::min(USB_MAX_SIZE, wbuf.len() - index);
+            self.usb_write_bulk(&wbuf[index..index + databytes])?;
+            index += databytes;
+        }
+
+        let mut resp = RspTransfer::new();
+        let bytecount = self.usb_read_bulk(&mut resp.as_bytes_mut())?;
+        ensure!(
+            bytecount >= 4,
+            Error::CommunicationError("Unrecognized response to I2C request")
+        );
+        ensure!(
+            resp.status_code == 0,
+            Error::CommunicationError("I2C error")
+        );
+        let databytes = bytecount - 4;
+        rbuf[0..databytes].clone_from_slice(&resp.data[0..databytes]);
+        let mut index = databytes;
+        while index < rbuf.len() {
+            let databytes = self.usb_read_bulk(&mut resp.data[index..])?;
+            ensure!(
+                databytes > 0,
+                Error::CommunicationError("Unrecognized reponse to I2C request")
+            );
+            index += databytes;
+        }
+        Ok(())
+    }
+
+    /// Send one USB packet.
+    fn usb_write_bulk(&self, buf: &[u8]) -> Result<()> {
+        self.inner
+            .usb_device
+            .borrow()
+            .write_bulk(self.interface.out_endpoint, buf)?;
+        Ok(())
+    }
+
+    /// Receive one USB packet.
+    fn usb_read_bulk(&self, buf: &mut [u8]) -> Result<usize> {
+        Ok(self
+            .inner
+            .usb_device
+            .borrow()
+            .read_bulk(self.interface.in_endpoint, buf)?)
+    }
+}
+
+impl Bus for HyperdebugI2cTarget {
+
+    fn run_transaction(&self, addr: u8, transaction: &mut [Transfer]) -> Result<()> {
+        let mut idx: usize = 0;
+        while idx < transaction.len() {
```

Comment:
I think this might also benefit from having each branch modify `transaction` to move the pointer forward.

Reply:
Nice suggestion, this makes the code a bit cleaner.